### PR TITLE
Reimplement external activities copying

### PR DESCRIPTION
**Pull request by pjanik**
@scytacki, I'm opening this PR so you can take a look at the code and it's easier for me to address your comments tomorrow. I'll fix and add tests tomorrow.

The main change is that now copying respects portal settings and it's simpler. It's following this [diagram](https://sequencediagram.org/index.html#initialData=C4S2BsFMAIDlIO7QMIHsAOBPaAZAggEp7QGQDOqArgE4DGMAZtagLbQAKq1wAhuAFDoe3ELRBCAdsGgAhZgjKRqg4aDGTpnbnxUj1PKbkJ5+prb3AAeALRzUCpQC5Ki6tFrhRAazLuMmfh5aUAA3HmAYcx0o8GsAPnwiRxjoCPBwX0TiYFQ-LFSACxgAHQkg0LBsLmhFAEdKSAl6Up4yCjFwyAATaAQwAsKYAFEADwjqMvBoPGCQEMqAOkDZsIijIn4AI1QR6FQQpXW8Ryy8kHJB6GpyKjpIBehS068Je18EAtzgIugXQ5BfKVNpBaDw-pd0FwLKlUOAemAHk9jNAJJBur4cqlqC5pN8YJDtFMWttKLiCgCljEbNYsidkdcyJCJF13v0IZRNp4yOSJABzR7M8I8aAMap4lGIcDYWgYc49Bm3ej8LqQcpzTpHLY7PYHNwxRyopDoajqtalUbjSYzCrATClWjXTo9fgxazxOwOaiOAnQxTM3zXLoga7BGEClhOsAisU-Q3QC1KK0rSrK1UrDUxIA). It also adds source_type property.

My general impression is that this new path is way easier to follow (and diagram :wink:), but the only drawback is that there's a new code path for publishing. Here it's not very problematic, it's a bit worse in LARA (I'll describe it in its own PR).

### CODE REVIEW

File: app/models/external_activity.rb [hunk 1]
```diff
@@ -207,6 +207,80 @@ def duplicateable?(user)
...
+    end
+
+    # Copy rest of the activity properties.
+    clone.material_property_list = material_property_list
+    clone.grade_level_list = grade_level_list
+    clone.subject_area_list = subject_area_list
+    clone.sensor_list = sensor_list
+    # Copy projects. Limit projects to ones that can be assigned by the new author.
+    allowed_projects = Admin::Project.all.select{ |p| Admin::ProjectPolicy.new(new_owner, p).assign_to_material? }.map(&:id)
+    clone.project_ids = project_ids & allowed_projects
```
Comment (scytacki) [inline app/models/external_activity.rb]: How about replacing `Admin::Project.all` with `Admin::Project.where(id: project_ids)`?

Then you can drop line 239 and we don't need to load in every project in the database.

Comment (pjanik) [reply app/models/external_activity.rb]: Thanks, good point, done.

**Comment by scytacki** [comment]
Looks good. Just needs some tests. WebMock can be used to mock the response from LARA.

**Comment by pjanik** [comment]
@scytacki, it's ready now. It took me a bit more than expected, as I haven't noticed that I was copying template reference before. It should work fine now, and tests are checking this case too.

**Comment by scytacki** [comment]
👍

**Pull request merged**